### PR TITLE
BUG: Segfaults due to invalid/out-of-bound sparse matrix constructor

### LINKED ISSUE #8778: csr_matrix can be constructed with out-of-bounds indices, leading to later segfaults
A `csr_matrix` can be constructed with malformed data (column indices that are out of bounds) without crashing. This can lead to later segfaults when (I guess) out of bounds memory is accessed. In my application, it didn't crash 100% of the time, but was more random, as is common with out-of-bounds memory accesses. I would expect the construction of the `csr_matrix` to fail, or at least the documentation should reflect that this can lead to undefined behavior if your parameters is incorrect.

### Reproducing code example:
```
from scipy.sparse import csr_matrix

data = [1.0, 1.0]
indices = [1001, 555]
indptr = [0, 1, 2]
shape = (2, 1000)

mat = csr_matrix((data, indices, indptr), shape=shape)
print "constructed csr_matrix"
print mat * mat.T

```

The seg fault occurs on the `print mat * mat.T` line, not the construction of the matrix with the malformed parameters.

### Error message:
```
$ python test.py
constructed csr_matrix
Segmentation fault (core dumped)
```

### Scipy/Numpy/Python version information:
```
('1.0.0', '1.14.0', sys.version_info(major=2, minor=7, micro=6, releaselevel='final', serial=0))
```
I'm running on Ubuntu 14.04 on a 64-bit x86 processor.

### PR DESCRIPTION
**Pull request by ewimberley**
Closes #8778
Closes #9253

Note that this has performance implications for loading sparse matrices using scipy.sparse.load_npz or when constructing sparse matrices using the tuple constructor. Invalid sparse matrices likely lead to a segfault, and are a security concern due to a "write-what-where" vulnerability in which an attacker could form a sparse matrix that reads/writes any address relative to the matrix object in memory. Because it uses the tuple version of the constructor, this can occur when load_npz is called to load a sparse matrix from a file.

It is very unlikely anyone would exploit this vulnerability due to the small number of potential victims and the small chance that someone would load a sparse matrix npz file from a source they did not trust. Therefore disclosure is not a big issue. However, it is a huge pain to debug this bug when encountered accidentally.

This commit provides a compromise between security/stability and speed, performing a full check of the sparse matrix in conditions where an error is likely to occur (creating a new matrix or loading one from disk), and not during matrix operations. Users who require performance for these operations can turn off full_check option using the safety_check argument in the constructor.

### CODE REVIEW

**Comment by ewimberley** [comment]
If this is too much of a performance penalty, safety_check=False can be set in methods like tocoo, tocsr, tocsc, and tobsr, transpose, getrow, getcol, etc. where the tuple constructor is used.

**Comment by ewimberley** [comment]
The sparse benchmarks show very little difference between these commits.
[performance_comparison.txt](https://github.com/scipy/scipy/files/6795612/performance_comparison.txt)

File: scipy/sparse/bsr.py [hunk 1]
```diff
@@ -223,7 +223,10 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False, blocksize=None):
         if dtype is not None:
             self.data = self.data.astype(dtype, copy=False)
 
-        self.check_format(full_check=False)
+        if isinstance(arg1, tuple):
+            self.check_format(full_check=safety_check)
```
Comment (perimosocordiae) [inline scipy/sparse/bsr.py:227]: This feels a little counterintuitive to me. It seems like we actually want to have 3 options for the safety check parameter: never, always, and "auto" (the default).

**Review by perimosocordiae — COMMENTED** [review]
I'm not sure if I'm on board with this addition yet. I think we could achieve a similar result by doing a better job of documenting the check_format method, without adding additional complexity to more constructors.

I'm interested to see what others think on this.

File: scipy/sparse/bsr.py (hunk 1 above)
Comment (ewimberley) [reply scipy/sparse/bsr.py:227]: @perimosocordiae Thanks for the feedback. I'm open to other solutions, or enhancements to this solution. When users encounter this bug they only get a segfault, and when sparse matrices are used by other libraries it can make the problem very difficult to debug. Documenting check_format would certainly help, but it does not prevent the undefined behavior from occurring unless users call it (leading to the same performance loss as the proposed solution). The reasons I went with the design of a flag in the constructor are as follows:

- It allows users to turn off the checks at their own risk
- The check only occurs once time, and not during matrix operations after it has been confirmed that the matrix is valid
- Checks only occur when the tuple constructor is used, which is the first opportunity to detect the error, and will only happen when loading a matrix from disk, or when advanced users are constructing a sparse matrix manually

Another solution I thought about is to perform the bounds check in the C++ code, mentioned in bug #8778, and while this would guarantee that no out of bounds reads/writes occur, it would have a significant performance impact.

Comment (ewimberley) [reply scipy/sparse/bsr.py]: I made the requested change to have a "never", "always", and "auto" safety check mode.

**Comment by nkaretnikov** [comment]
@perimosocordiae hello, i've looked at this as part of investigating issue #9253 (and other related ones).

general thoughts:
- i think scipy should strive to be "correct by construction" as incorrectly constructed things can lead to problems later, which might be hard to catch (could lead to crashes or even incorrect results). so i would suggest checking the format in the constructors.
- if checking the format in the constructors is not possible due to a perf penalty (even tho the benchmark above doesn't seem to support this claim), i would still suggest adding a parameter to perform the check in each constructor, but disabling it by default.
- there could also be a global flag, similar to a strict mode in other languages, that would allow users to opt-in for more (potentially slower) checking without changing their code. the only disadvantage here is that it might be hard to know which mode is on if a global is used for this.

about the patch:
- as mentioned above, i like the idea of checking things in the constructors.
- in this patch, i dislike the if/then logic related to checking. i think the checking parameter passed to the constructors should go directly to `check_format` without additional logic relying on shapes. this will make it more transparent for users whether format checking is used or not.
- in this patch, i also dislike the hardcoded string parameters. i think these should be defined as constants/enums somewhere and should be called similar to how they are called in `check_format`. i'd suggest calling the argument `check_format` with this range of expected values: `no_check`, `basic_check`, and `full_check` (or the same without the `_check` part as it seems redundant), defined as constants/or an enum. any other value should error out with an error message.
- the new parameter needs to be documented.

i'm planning to refine this patch to demonstrate this approach, stay tuned!

**Comment by rgommers** [comment]
> I think scipy should strive to be "correct by construction"

I agree - the default should be to check. And only if there's a significant performance penalty should we consider adding a keyword (not pretty) to bypass the check.


> if checking the format in the constructors is not possible due to a perf penalty (even tho the benchmark above doesn't seem to support this claim),

Yes indeed, all close to 1.0 before/after. Let's see if there are cases where it actually matters. Then I'd still try moving the check to C++ or Cython and always doing it to avoid adding the keyword.

**Comment by nkaretnikov** [comment]
yes, i agree with the idea of not introducing a keyword. once introduced, it'll be hard to remove later and this seems like an implementation detail.

also, i'll mention here for the sake of visibility: there are other issues with this patch/approach, which i'm going to fix:
- inconsistent naming (and different implementations) of shape-checking functions for different matrix types. for instance, `check_format` for csr and bsr, `_check` for coo (without an argument: full check is always performed).
- a major issue: in the `check_format` ones, `self.prune()` is used before performing the full check, which relies on `self.nnz` being set. i know for sure that `self.nnz` for csr it's not set in the constructor, which means that in `self.prune`, `self.indices` and `self.data` will be sliced to contain zero elements. there's also a check later in `check_format` relying on `self.nnz` being set, so it must be set for the full check to be performed. the check exists to avoid erroring out on min/max operations if i understand correctly.

**Comment by nkaretnikov** [comment]
tl;dr: i think it's better to focus on fixing specific crashes on the c/cpp side.

more background:

i created a small proof of concept/work in progress branch here:
https://github.com/nkaretnikov/scipy/commits/check-matrix-format-9253

it passes all csr tests but fails quite a few base tests, mainly because of inplace operations/data consistency issues during those.

there, i tried to push further the idea of using `check_format`, specifically, after every update to critical structures of a csr instance. my thought process was: if you do it just in the constructor (like in this pr) or with some operations (which is error-prone as you can miss some), nothing prevents someone to modify `indptr` (or other critical attributes later) to cause a crash. so i gathered i should just try hooking all the updates. including updates to mutable attribute structures, like `self.indptr[1] = 42`.

however, this approach turned out to be problematic for the following reason: there are many internal methods in scipy that modify `indptr` (and other critical attributes) while performing inplace operations. so in those types of methods, the internal consistency of a csr instance is invalidated temporarily, but there's no way for `check_format` to pass in those cases because it's called before the consistency is restored. you could potentially create a new instance in every case like that (by going thru a constructor) but that seems impractical/costly. it's just too much code/bloat. this approach also has this problem that now you're using this custom wrapper class (`_checked_array`) instead of just using pure numpy arrays. so seems like a dead end.

another thing i've tired is just freezing ndarrays (used for `indptr` and such) by disabling writes:
https://numpy.org/doc/stable/reference/generated/numpy.ndarray.setflags.html

but this approach has two issues:
- if you allow enabling writes later, this only moves the original problem somewhere else (the code might still crash)
- if you disallow enabling writes later, this breaks inplace operations (and you don't want to do copies)
- there's actually no reliable way of doing `write=False` followed by `write=True` somewhere later because in some cases the relevant memory might be owned by something else, so you cannot re-enabled the write property.

i think the way to go here is this:
- write fail-safe code on c/cpp side such that it doesn't access invalid memory (assume malicious data is passed to functions), error out on invalid data (instead of crashing). this seems to be the only practical choice with the current design despite requiring to hunt for these problematic cases. this will probably also lead to error messages being less user-friendly, but it's okay. at least, this way we don't pay the perf penalty and this only affects small amount of code -- so easier to test.
- in some cases, to provide more user-friendly error messages earlier, `check_format` calls could be inserted in some places, but the assumption should be that all code should be able to handle all input everywhere.

**Comment by perimosocordiae** [comment]
I agree with @nkaretnikov 's analysis: we're unfortunately in a situation where direct modification of spmatrix internals is commonplace, and no amount of defensive coding or checking will be able to prevent malformed inputs from making their way into the compiled helper functions.

Let's close this PR and follow up with gh-16243 and similar efforts to avoid segfaults from the C++ side.
